drake/systems/primitives/random_source.h

```diff
@@ -26,48 +49,53 @@ namespace systems {
 /// necessary information for stochastic analysis).
 ///
 /// @ingroup primitive_systems
-template <typename Distribution>
+template <typename Distribution, typename Generator = std::mt19937>
 class RandomSource : public LeafSystem<double> {
  public:
   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(RandomSource)
 
+  typedef systems::RandomState<Distribution, Generator> RandomState;
+  typedef typename RandomState::Seed Seed;
+
   /// Constructs the RandomSource system.
   /// @param num_outputs The dimension of the (single) vector output port.
   /// @param sampling_interval_sec The sampling interval in seconds.
   RandomSource(int num_outputs, double sampling_interval_sec) {
-    this->DeclareDiscreteUpdatePeriodSec(sampling_interval_sec);
+    this->DeclarePeriodicUnrestrictedUpdate(sampling_interval_sec, 0.);
     this->DeclareVectorOutputPort(BasicVector<double>(num_outputs),
                                   &RandomSource::CopyStateToOutput);
     this->DeclareDiscreteState(num_outputs);
+    this->DeclareAbstractState(AbstractValue::Make(RandomState(seed_)));
   }
 
   /// Initializes the random number generator.
-  void set_random_seed(double seed) { generator_.seed(seed); }
+  void set_random_seed(Seed seed) { seed_ = seed; }
 
  private:
   // Computes a random number and stores it in the discrete state.
-  void DoCalcDiscreteVariableUpdates(
-      const drake::systems::Context<double>& context,
-      drake::systems::DiscreteValues<double>* updates) const override {
+  void DoCalcUnrestrictedUpdate(const Context<double>& context,
+                                State<double>* state) const override {
+    auto& random_state =
+        state->template get_mutable_abstract_state<RandomState>(0);
+    auto* updates = state->get_mutable_discrete_state();
     const int N = updates->size();
     for (int i = 0; i < N; i++) {
-      double random_value = distribution_(generator_);
-      (*updates)[i] = random_value;
+      (*updates)[i] = random_state.GetNextValue();
     }
   }
 
+  std::unique_ptr<AbstractValues> AllocateAbstractState() const override {
+    return std::make_unique<AbstractValues>(
+        AbstractValue::Make(RandomState(seed_)));
+  }
+
   // Output is the zero-order hold of the discrete state.
   void CopyStateToOutput(const Context<double>& context,
                          BasicVector<double>* output) const {
     output->SetFromVector(context.get_discrete_state(0)->CopyToVector());
   }
 
-  // Note: currently there is undeclared state in the variables below.
-  // TODO(russt): Use abstract state to save the parameters of the generator and
-  // distribution (waiting on event scheduling for abstract states).
-  // TODO(russt): Obtain consistent results across multiple platforms (#4361).
```

Comment:
Yessir, I had moved line 68 to [line 29](https://github.com/RobotLocomotion/drake/pull/6279/files#diff-200ea6f0400992a71e7520bfcc8c8a5eR29) in `RandomState`.

Would you like me to take on this `TODO` item, and assign #4361 to myself?